smartcontract/storage/cachedb.go

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (C) 2018 The ontology Authors
+ * This file is part of The ontology library.
+ *
+ * The ontology is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ontology is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package storage
+
+import (
+	comm "github.com/ontio/ontology/common"
+	"github.com/ontio/ontology/core/payload"
+	"github.com/ontio/ontology/core/store/common"
+	"github.com/ontio/ontology/core/store/overlaydb"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+// CacheDB is smart contract execute cache, it contain transaction cache and block cache
+// When smart contract execute finish, need to commit transaction cache to block cache
+type CacheDB struct {
+	memdb      *overlaydb.MemDB
+	backend    *overlaydb.OverlayDB
+	keyScratch []byte
+}
+
+// NewCacheDB return a new contract cache
+func NewCacheDB(store *overlaydb.OverlayDB) *CacheDB {
+	return &CacheDB{
+		backend: store,
+		memdb:   overlaydb.NewMemDB(0),
+	}
+}
+
+func ensureBuffer(b []byte, n int) []byte {
+	if cap(b) < n {
+		return make([]byte, n)
+	}
+	return b[:n]
+}
+
+func makePrefixedKey(dst []byte, prefix byte, key []byte) []byte {
+	dst = ensureBuffer(dst, len(key)+1)
+	dst[0] = prefix
+	copy(dst[1:], key)
+	return dst
+}
+
+// Commit current transaction cache to block cache
+func (self *CacheDB) Commit() {
+	self.memdb.ForEach(func(key, val []byte) {
+		if len(val) == 0 {
+			self.backend.Delete(key)
+		} else {
+			self.backend.Put(key, val)
+		}
+	})
+}
+
+//
+//// Add item to cache
+//func (self *CacheDB) Add(prefix common.DataEntryPrefix, key []byte, value states.StateValue) {
+//	data := new(bytes.Buffer)
+//	// todo: remove error or return to caller
+//	err := value.Serialize(data)
+//	if err != nil {
+//		panic(err)
+//	}
+//
+//	self.keyScratch = makePrefixedKey(self.keyScratch, byte(prefix), key)
+//	self.memdb.Put(self.keyScratch, data.Bytes())
+//}
+
+func (self *CacheDB) Put(key []byte, value []byte) {
```

Comment:
why not use StorageItem as second param?

Reply:
it can be any serialized structure defined in contract. not just `StorageItem`. Like:
```
FooContract.go:
type Person struct {
	Version byte
	Name string
	Age int
}
func (self *Person) ToArray() []byte{...}

cache.Put([]byte("p1"), person1.ToArray())
```